atop, netatop, nixos/atop: improve packaging and options

pschyska opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
The atop package where missing some files from upstream that I wanted to use (systemd units), mostly because `$out` instead of `$(out)` was used in a make context and this led to them not being copied.
The netatop package was missing the unit file for the netatop service, which is required for process accounting.
The nixos/atop module was extended to include configuration for all of atop's standard components, and optionally atopgpu and netatop. 

Because I don't have a NVIDIA GPU, I wasn't able to test `atopgpud`.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests)) (added new test)
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

pschyska committed:
netatop: package netatop.service

pschyska committed:
atop: fix packaging .service files, add option to build with atopgpu

pschyska committed:
nixos/atop: Add configuration for atop services, allow to enable netatop, gpuatop, allow setuid wrapper

r-rmcgibbo commented:
Result of `nixpkgs-review pr 123053` at a8898703 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxPackages-libre.netatop</li>
    <li>linuxPackages_4_4.netatop</li>
    <li>linuxPackages_4_9.netatop</li>
    <li>linuxPackages_hardkernel_4_14.netatop</li>
    <li>linuxPackages_hardkernel_latest.netatop</li>
    <li>linuxPackages_latest-libre.netatop</li>
    <li>linuxPackages_testing_bcachefs.netatop</li>
    <li>linuxPackages_xanmod.netatop</li>
  </ul>
</details>
<details>
  <summary>11 packages built successfully:</summary>
  <ul>
    <li>atop</li>
    <li>linuxPackages.netatop (linuxPackages_5_10.netatop)</li>
    <li>linuxPackages_4_14.netatop</li>
    <li>linuxPackages_4_19.netatop</li>
    <li>linuxPackages_5_11.netatop (linuxPackages_latest.netatop)</li>
    <li>linuxPackages_5_12.netatop</li>
    <li>linuxPackages_5_4.netatop</li>
    <li>linuxPackages_hardened.netatop</li>
    <li>linuxPackages_latest_hardened.netatop</li>
    <li>linuxPackages_lqx.netatop</li>
    <li>linuxPackages_zen.netatop</li>
  </ul>
</details>

7c6f434c commented on nixos/tests/atop.nix:
```diff
@@ -0,0 +1,132 @@
+      a_atopgpu(m, atopgpu)
+      a_atoprc(m, atoprc)
+
+    assert_all(defaults, False, True, True, True, False, False, False)
```
Have you considered `with_subtests` like in `installer.nix` and decided against it, or did you not know about its existence?

pschyska replied:
I didn't really know about it. I was unhappy with the tests anyways, and I've rewritten them just now. Is this any better?

7c6f434c replied:
Looks fine to me. But if you think it's better now, that's already enough for me.

pschyska committed:
nixos/atop: Never enable setuidWrapper by default, rename service/timer enabling options

pschyska committed:
nixos/atop: Rework the test

- use "with subtest" everywhere
- do more in nix and less in python
- use makeTest directly to define multiple tests instead of one with
  multiple nodes -> this enables them to run in parallel

7c6f434c commented:
@ofborg test atop.everything

7c6f434c commented on pkgs/os-specific/linux/atop/default.nix:
```diff
@@ -9,31 +19,52 @@ stdenv.mkDerivation rec {
   ];
 
+  patches = [
+    ./atop-pm.sh.patch
```
I guess I agree with the auto-linting that adding a bunch of patches would be clearer with a comment in the place of use

pschyska replied:
I have shuffled around the patches into fix-paths.patch for atop and netatop each (replacing FHS references with @@ templates), and commented the rest

pschyska committed:
nixos/atop: Add a note about netatop tainting the kernel

pschyska committed:
nixos/atop: Split up restart triggers between atop and netatop

pschyska committed:
netatop: Split out path fixing and PIDFile changes to separate patches and comment

pschyska committed:
atop: Merge all path-fixing to a single patch, comment the remaining patches

pschyska commented:
@ofborg test atop.everything

pschyska commented:
Sorry @7c6f434c, I accidentally requested a re-review just now.

7c6f434c commented:
@ofborg test atop.default

7c6f434c commented:
@ofborg test atop.defaults

I am not a very careful person

7c6f434c reviewed (approved):
Diff looks good

SuperSandro2000 commented on pkgs/os-specific/linux/netatop/netatop.service.patch:
```diff
@@ -0,0 +1,7 @@
++++ b/netatop.service
+@@ -11,3 +11,3 @@
+ ExecStopPost=@kmod@/bin/rmmod netatop
+-PIDFile=/var/run/netatop.pid
++PIDFile=/run/netatop.pid
```
Those should be symlinked to one another.

pschyska replied:
Yes, they are, at least on my system. However systemd warns about using /var/run and I wanted to suppress that warning.
I could live with the warning and remove that patch though, if you insist :-)

aanderse replied:
Bonus points if you submit a patch upstream instead. Many distros warn about this by now. Fix it upstream: fix it everywhere.

pschyska replied:
`/var/run -> /run` is actually not true in all instances. Atop supports many distros. The binaries do write to /var/run (hardcoded AFAICT). So changing it upstream breaks everywhere these are not symlinked.
In the end, it doesn't really matter, so I suggest to either leave it as is, and live with the systemd warning, or patch it like proposed.

aanderse replied:
@pschyska thanks for letting me know that. Debian switched back in 2013 so I figured pretty much every distro must have switched over by now without ever having verified how valid this assumption was. I am curious which distros have not. Do you have any links I can read up on?

Back on topic... I'm not sure what to say here. I guess the patching is fine then. What do you think @SuperSandro2000?

pschyska replied:
@aanderse Nothing to read up on, just a gut feeling that upstream wouldn't take the patch because it can't be assumed that these are symlinked in general.

aanderse replied:
Oh. I thought you were saying there were some concrete example distros of this. In that case... even Debian switched back in 2013, and this is part of the FHS. Your gut feeling certainly could be correct, but what isn't correct would be software fighting against this change. :disappointed: 

> In FHS 3.0, /var/run is replaced by /run

SuperSandro2000 replied:
Could also be a substituteInPlace but a patch is also fine.

pschyska replied:
@aanderse  https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s13.html "It is *valid* to implement /var/run as a symlink to /run." (emphasis mine). That said, `/var/run` is still defined to exist as to ` ensure compatibility with systems and software using an older version of this specification.`. So it's still "legal" to use `/var/run`, yet systemd's gonna systemd and decided to annoy the user with endless warnings ;-)

SuperSandro2000 commented on pkgs/os-specific/linux/atop/default.nix:
```diff
@@ -9,31 +19,55 @@ stdenv.mkDerivation rec {
+  '' + (if withAtopgpu then ''
+    wrapPythonPrograms
+  '' else ''
+    rm $out/lib/systemd/system/atopgpu.service
+    rm $out/bin/atopgpud
+    rm $out/share/man/man8/atopgpud.8
```
```suggestion
    rm $out/lib/systemd/system/atopgpu.service $out/bin/atopgpud $out/share/man/man8/atopgpud.8
```

7c6f434c replied:
Not sure why introduce a line longer than 80 characters here.

SuperSandro2000 replied:
nix does not care about 80 line length but whatever

SuperSandro2000 commented on pkgs/os-specific/linux/atop/default.nix:
```diff
@@ -9,31 +19,55 @@ stdenv.mkDerivation rec {
   '';
+  postInstall = ''
+    # remove extra files we don't need
+    rm -rf $out/{var,etc}
+    rm -rf $out/bin/atop{sar,}-${version}
```
```suggestion
    rm -r $out/{var,etc} $out/bin/atop{sar,}-${version}
```

SuperSandro2000 commented on pkgs/os-specific/linux/atop/default.nix:
```diff
@@ -9,31 +19,55 @@ stdenv.mkDerivation rec {
+    done
+
     sed -e 's/chown/true/g' -i Makefile
-    sed -e '/chkconfig/d' -i Makefile
     sed -e 's/chmod 04711/chmod 0711/g' -i Makefile
```
Please use substituteInPlace.

pschyska replied:
Okay, I didn't change it because it was like that in master.

I did however already have problems with the indiscriminate replacing business, so I'd rather add a patch for Makefile, if that works for you? I'll still use substitute for dynamic values (out, paths to dependencies, etc.) but for the static modifications it doesn't make much sense in the first place (e.g. disabling the chown by replacing it with "true")

Does that work for you?

SuperSandro2000 replied:
substituteInPlace should be used for simple replaces instead of patches to keep the repository small.

SuperSandro2000 commented on pkgs/os-specific/linux/atop/default.nix:
```diff
@@ -9,31 +19,55 @@ stdenv.mkDerivation rec {
-    "DEFPATH=$out/etc/default"
-    "ROTPATH=$out/etc/logrotate.d"
+    "DESTDIR=$(out)"
+    "BINPATH=/bin"
+    "SBINPATH=/bin"
+    "MAN1PATH=/share/man/man1"
+    "MAN5PATH=/share/man/man5"
+    "MAN8PATH=/share/man/man8"
+    "SYSDPATH=/lib/systemd/system"
+    "PMPATHD=/lib/systemd/system-sleep"
```
Why do they not include $out?

pschyska replied:
They are all used as like this: `$(DESTDIR)$(BINPATH)`, so `$out` is correctly prependet.

pschyska replied:
@SuperSandro2000 Works for you? Alternativeley I can `DESTDIR=""` and fully qualify all the names, but then I'd have to go through the Makefile again because I only mentioned the keys that were relevant, and were wrong by default with `DESTDIR=$(out)`.

SuperSandro2000 commented on pkgs/os-specific/linux/atop/default.nix:
```diff
@@ -9,31 +19,55 @@ stdenv.mkDerivation rec {
   };
 
-  buildInputs = [zlib ncurses];
+  nativeBuildInputs = if withAtopgpu then [ python3.pkgs.wrapPython ] else [ ];
+  buildInputs = [ zlib ncurses ] ++ (if withAtopgpu then [ python3 ] else [ ]);
+  pythonPath = if withAtopgpu then [ python3.pkgs.pynvml ] else [ ];
```
```suggestion
  nativeBuildInputs = lib.optionals withAtopgpu [ python3.pkgs.wrapPython ];
  buildInputs = [ zlib ncurses ] ++ lib.optionals withAtopgpu [ python3 ];
  pythonPath = lib.optionals withAtopgpu [ python3.pkgs.pynvml ];
```

SuperSandro2000 commented on pkgs/os-specific/linux/atop/default.nix:
```diff
@@ -1,4 +1,14 @@
+, findutils
+, systemd
+, python3
+  # makes the package unfree via pynvml
```
```suggestion
# makes the package unfree via pynvml
```

SuperSandro2000 commented on pkgs/os-specific/linux/atop/atopacct.service.patch:
```diff
@@ -0,0 +1,7 @@
++++ b/atopacct.service
+@@ -9,3 +9,3 @@
+ Type=forking
+-PIDFile=/var/run/atopacctd.pid
++PIDFile=/run/atopacctd.pid
```
Those should be symlinked.

pschyska replied:
see https://github.com/NixOS/nixpkgs/pull/123053#discussion_r633151414

pschyska committed:
atop:, nixos/atop: Apply style suggestions

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

pschyska committed:
atop: use substituteInPlace instead of sed

pschyska commented:
❤️ Thank you @7c6f434c @aanderse @SuperSandro2000

samueldr commented on nixos/modules/programs/atop.nix:
```diff
@@ -12,11 +12,83 @@ in
+        };
+        package = mkOption {
+          type = types.package;
+          default = config.boot.kernelPackages.netatop;
```
`detaultText` attribute missing for a package.

This can lead to breakage in some evals:

```
error: attribute 'netatop' missing, at /nix/store/md569yk7k26jdknlbarfyai4i1lb7kkj-nixexprs.tar.xz/nixos/modules/programs/atop.nix:38:21
```

pschyska replied:
Hey @samueldr, sorry I broke your website build :-(
I'm opening a PR momentarily adding defaultTexts.
I'm having a problem with the tests locally, but I think I've got a fix already.

pschyska replied:
https://github.com/NixOS/nixpkgs/pull/123926

samueldr replied:
No biggie about the build failure :). It's an unusual failure type that, for this particular package, won't be a problem on the Hydra for NixOS itself. So it's really not obvious that it can cause some fringe breakage.

Shados commented:
This rather unexpected broke my atoprc because `programs.atop.enable` did not previously exist, but is now required to build it. Given this is a breaking change, it should probably *at least* get a mention in the NixOS release notes, if not an automated warning (although given the nature of the change, I don't think that could be cleanly done).

pschyska commented:
@Shados Sorry for breaking your config. This case slipped my attention. I just opened a mini PR adding a test for your case and a fix.

FRidh commented:
Note this allows the use of unfree packages. I am going to revert that part.
